docs/designs/api-design.md

```diff
@@ -0,0 +1,343 @@
+the newly created program.
+
+## Get Program
+
+Get details for the Program page.
+
+`GET /program/<programId>`
+
+**Request**
+
+```json
+
+```
+
+**Response**
+
+```json
+{
+  "program": {
+    "id": "<uuid>",
+    "name": "<string>",
+    "description": "<string>",
+
+    // _internal metadata used by the backend
+    "_internal": {
+        "lastUpdatedTime": "<iso-date-str>",
+        "lastUpdatedUser": "<userId>",
+    },
+
+    "projects": [
+        {
+            "id": "<uuid>",
+            "name": "<string>",
+            "description": "<string>",
+            "status": "<status>",
+            "tasks": [
+                {
+                    "id": "<uuid>",
+                    "name": "<string>",
+                    "description": "<string>",
+                    "status": "<string>",
+                    "assignee": [
+                        // List of users assigned to this task
+                        {
+                            "id": "<uuid>",
+                            "firstName": "<string>",
+                            "thumbnailColor": "<hexcode>"
+                        }
+                    ],
+                    "budget": {
+                        "amount": "<number>",
+                        "currency": "<string>",
+                    },
+                    "duration": "<string>",
+                    "estimatedStartDate": "<iso-date-str>",
+                    "estimatedEndDate": "<iso-date-str>",
+                    "actualStartDate": "<iso-date-str>",
+                    "actualEndDate": "<iso-date-str>"
+                }
+            ],
+        }
+    ]
+  },
+
+  // All information necessary for visual presentation is in this object
+  "ui": {
+      "programs": {
+        "<programId>": {
+            "projectsOrder": ["projectId1", "projectId2" ...]
+        }
+      },
+      "projects": {
+          "<projectId>": {
+            "taskOrder": ["taskId1", "taskId2" ...]
+          }
+      }
+    }
+}
+```
+
+## Update Program: Add/Update Project, Add/Update Task Overview
+
+Called to update an existing program, details about its constituent projects, and high level details about
+individual tasks within the project. It is also used when you add a new project or a new task. It is a very
+broadly scoped API overloading a lot of functionality into one. We are using this structure for now because it
+simplifies the frontend UI logic.
+
+`PUT /program/<programId>`
+
+**Request**
+Request sends an update copy of the entire structure returned by `Get Program API`. Backend will take care of splicing
+individual elements and storing them in appropriate tables.
+
+**Response**
+Returns a complete response of `Get Program API`. Makes it very easy for frontend to refresh the program page using
+the existing Program page rendering logic.
+
+**HTTP Status Code**
+
+- `409 Conflict`: This is returned when new project cannot be created because the program state stored in the database
```

Comment:
Let's consider using `If-Match` or `If-Unmodified-Since` and return `412` for detecting these types of conflicts https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/412#avoiding_mid-air_collisions